### PR TITLE
enhancement: add relation data to history versions response

### PR DESCRIPTION
**Pull request by remidej**
**_ℹ️  This PR replaces #19786, as the git history became too much of a mess on that one_**

### What does it do?

- Saves only the ID in the data key when history versions are related
- On retrieval, for each relation and media, populate an object with results and missingCount
- This PR starts messing around with the frontend too (relation input in history), but it's mostly pseudocode to make sure the approach works. The actual implementation is a separate ticket.

### Why is it needed?

We need to display relations in history. And showing that some relations were deleted requires special handling.

### How to test it?

🚨 Because of the data structure change, your app may crash with pre-existing history versions 🚨

Create history versions with some relations. View history, you should see the data. Go delete the related content, you should see that too in the history frontend.

### CODE REVIEW

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 1]
```diff
@@ -177,25 +264,124 @@ const createHistoryService = ({ strapi }: { strapi: Core.Strapi }) => {
...
       ]);
 
-      const versionsWithMeta = results.map((version) => {
-        return {
-          ...version,
-          meta: {
-            unknownAttributes: getSchemaAttributesDiff(
-              version.schema,
-              strapi.getModel(params.contentType).attributes
-            ),
```
Comment (markkaylor) [inline packages/core/content-manager/server/src/history/services/history.ts:187]: Did you mean to remove this? I don't see it moved anywhere in the diff and we still need it.

File: packages/core/content-manager/shared/contracts/history-versions.ts [hunk 2]
```diff
@@ -31,12 +32,6 @@ export interface HistoryVersionDataResponse extends Omit<CreateHistoryVersion, '
     email: string;
   };
   locale: Locale | null;
-  meta: {
-    unknownAttributes: {
-      added: Struct.SchemaAttributes;
-      removed: Struct.SchemaAttributes;
-    };
-  };
```
Comment (markkaylor) [inline packages/core/content-manager/shared/contracts/history-versions.ts:39]: Same comment as above, I think we still need this in the response.

**Review by markkaylor — COMMENTED** [review]
Couple quick questions before I give it a deeper look and test it out

File: packages/core/content-manager/server/src/history/services/history.ts (hunk 1 above)
Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:187]: ah shit didn't mean to remove it, just move it. thx

**Comment by derrickmehaffy** [comment]
This looks interesting, could you share a quick screenshot of what this looks like?

**Comment by remidej** [comment]
@derrickmehaffy the frontend isn't ready but it will look like this
CleanShot 2024-04-04 at 12 57 52@2x

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 3]
```diff
@@ -177,25 +265,130 @@ const createHistoryService = ({ strapi }: { strapi: Core.Strapi }) => {
...
+        | null;
+
+      /**
+       * Get an object with two keys:
+       * - results: an array with the current values of the relations
+       * - meta: an object with the count of missing relations
+       */
+      const buildRelationReponse = async (
+        values: EntryToPopulate[],
+        attributeSchema: Struct.SchemaAttributes[keyof Struct.SchemaAttributes]
```
Comment (markkaylor) [inline packages/core/content-manager/server/src/history/services/history.ts:283]: Is it not the same as `Schema.Attribute.AnyAttribute`?

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 4]
```diff
@@ -177,25 +265,130 @@ const createHistoryService = ({ strapi }: { strapi: Core.Strapi }) => {
...
+                 * or a normal relation. The extra checks are only for type narrowing
+                 */
+                let relatedEntry;
+                if (isNormalRelation) {
+                  if ('documentId' in entry) {
+                    relatedEntry = await strapi
+                      .documents(attributeSchema.target)
+                      .findOne(entry.documentId, { locale: entry.locale || undefined });
+                  }
+                } else if ('id' in entry) {
```
Comment (markkaylor) [inline packages/core/content-manager/server/src/history/services/history.ts:314]: what's the difference here? 🤔

File: packages/core/content-manager/shared/contracts/history-versions.ts [hunk 5]
```diff
@@ -12,7 +12,8 @@ export interface CreateHistoryVersion {
   locale: string | null;
   status: 'draft' | 'published' | 'modified' | null;
   data: Record<string, unknown>;
-  schema: Struct.SchemaAttributes;
+  schema: Schema.Attributes;
```
Comment (markkaylor) [inline packages/core/content-manager/shared/contracts/history-versions.ts:15]: Did you understand the difference between this and Struct.SchemaAttributes? I thought this one required a generic UID argument

**Review by markkaylor — APPROVED** [review]
Nice, I tested it kinda quickly but it looks like it's working. Left some comments but I think we are good to merge.

File: packages/core/content-manager/shared/contracts/history-versions.ts (hunk 5 above)
Comment (remidej) [reply packages/core/content-manager/shared/contracts/history-versions.ts:15]: you're right, didn't mean to overwrite that

File: packages/core/content-manager/server/src/history/services/history.ts (hunk 3 above)
Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:283]: i couldn't find the proper import, thanks!

File: packages/core/content-manager/server/src/history/services/history.ts (hunk 4 above)
Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:314]: documentId is needed for relations, id is needed for media assets. I added a comment

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy the frontend isn't ready but it will look like this
> CleanShot 2024-04-04 at 12 57 52@2x
> 

Looks awesome thank you 😊

**Pull request merged**